source/grid/tria_description.cc

```diff
@@ -0,0 +1,662 @@
+                }
+            }
+
+          const_cast<dealii::Triangulation<dim, spacedim> &>(tria)
+            .load_user_flags(old_user_flags);
+        }
+
+      return construction_data;
+    }
+
+
+
+    template <int dim, int spacedim>
+    Description<dim, spacedim>
+    create_description_from_triangulation_in_groups(
+      const std::function<void(dealii::Triangulation<dim, spacedim> &)>
+        &                                            serial_grid_generator,
+      const std::function<void(dealii::Triangulation<dim, spacedim> &,
+                               const MPI_Comm,
+                               const unsigned int)> &serial_grid_partitioner,
+      const MPI_Comm                                 comm,
+      const int                                      group_size,
+      const bool construct_multilevel_hierarchy)
+    {
+#ifndef DEAL_II_WITH_MPI
+      (void)serial_grid_generator;
+      (void)serial_grid_partitioner;
+      (void)comm;
+      (void)group_size;
+      (void)construct_multilevel_hierarchy;
+
+      return Description<dim, spacedim>();
+#else
+      const unsigned int my_rank =
+        dealii::Utilities::MPI::this_mpi_process(comm);
+      const unsigned int group_root = (my_rank / group_size) * group_size;
+
+      const int mpi_tag =
+        dealii::Utilities::MPI::internal::Tags::fully_distributed_create;
+
+      // check if process is root of the group
+      if (my_rank == group_root)
+        {
+          // Step 1: create serial triangulation
+          dealii::Triangulation<dim, spacedim> tria(
+            construct_multilevel_hierarchy ?
+              dealii::Triangulation<dim, spacedim>::none :
+              dealii::Triangulation<dim, spacedim>::
+                limit_level_difference_at_vertices);
+          serial_grid_generator(tria);
+
+          // Step 2: partition active cells and ...
+          serial_grid_partitioner(tria, comm, group_size);
+
+          // ... cells on the levels if multigrid is required
+          if (construct_multilevel_hierarchy)
+            GridTools::partition_multigrid_levels(tria);
+
+          const unsigned int end_group =
+            std::min(group_root + group_size,
+                     dealii::Utilities::MPI::n_mpi_processes(comm));
+
+          // 3) create ConstructionData for the other processes in group
+          for (unsigned int other_rank = group_root + 1; other_rank < end_group;
+               other_rank++)
+            {
+              // 3a) create construction data for other ranks
+              const auto construction_data =
+                create_description_from_triangulation(
+                  tria, comm, construct_multilevel_hierarchy, other_rank);
+              // 3b) pack
+              std::vector<char> buffer;
+              dealii::Utilities::pack(construction_data, buffer, false);
+
+              // 3c) send ConstructionData
+              const auto ierr = MPI_Send(buffer.data(),
```

Comment:
For a follow-up patch: You could be using an Isend here if you pack the data into a buffer that is part of an array that lives outside the loop.

Reply:
This is done intentionally. I only want to create the next `Description` once the current one has been sent away (to reduce memory consumption). I have added a comment.

Reply:
Hm, is this really an issue? I don't think it's particularly important, but suspect that matrices, vectors, and all other objects take far more memory than you might be using here -- but then they also take more time, so I don't think that it matters much.